promoter/database.go

```diff
@@ -0,0 +1,132 @@
+package promoter
+
+import (
+	"context"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"go.sia.tech/siad/crypto"
+
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/mongo/readconcern"
+	"go.mongodb.org/mongo-driver/mongo/readpref"
+	"go.mongodb.org/mongo-driver/mongo/writeconcern"
+)
+
+const (
+	dbName                  = "siacoinpromoter"
+	colWatchedAddressesName = "watchedaddresses"
+)
+
+type (
+	// updateFunc is the type of a function that can be used as a callback
+	// in threadedAddressWatcher.
+	updateFunc func(WatchedAddressesUpdate)
+
+	// WatchedAddress describes an entry in the watched address collection.
+	WatchedAddress struct {
+		// Address is the actual address we track. We make that the _id
+		// of the object since the addresses should be indexed and
+		// unique anyway.
+		Address crypto.Hash `bson:"_id"`
```

Comment:
nit: It's a bit confusing that the `_id` is not really a Mongo id. This is something the developer will need to keep in mind any time they work with this collection. Are you sure this is a good tradeoff?

Reply:
So the main reason for this apart from the address being a great candidate for an ID since it has the same properties is the way `Watch` works. By default it only provides you with the ID of the entry and the changes that happened to it. So in case of a deletion, we only get the ID and nothing else since the change was that the document is now gone.

I could try and specify `FullDocument` on `Watch` which will return the full document alongside the ID but I'm not sure if the `FullDocument` after a change will contain the address. I'd have to try it out.

Reply:
I see. It would be great if `FullDocument` works out because that would remove a nice "gotcha" from the codebase.

Reply:
I tried it but unfortunately the `delete` just doesn't contain the `fullDocument` even with changing the option from the default to the one that includes the whole document.